sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2WriteRequirements.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.sql.{catalyst, AnalysisException}
+import org.apache.spark.sql.catalyst.analysis.Resolver
+import org.apache.spark.sql.catalyst.expressions.SortOrder
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, RepartitionByExpression, Sort}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.distributions.{ClusteredDistribution, OrderedDistribution, UnspecifiedDistribution}
+import org.apache.spark.sql.connector.expressions.{Expression, FieldReference, NullOrdering, SortDirection, SortValue}
+import org.apache.spark.sql.connector.write.RequiresDistributionAndOrdering
+import org.apache.spark.sql.internal.SQLConf
+
+// this rule does resolution in the optimizer because some nodes like OverwriteByExpression
+// must undergo the expression optimization before we can construct a logical write
+object V2WriteRequirements extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
+    case V2BatchWriteCommand(write: RequiresDistributionAndOrdering, query, false) =>
+      val sqlConf = SQLConf.get
+      val resolver = sqlConf.resolver
+
+      val distribution = write.requiredDistribution match {
+        case d: OrderedDistribution =>
+          d.ordering.map(e => toCatalyst(e, query, resolver))
+        case d: ClusteredDistribution =>
+          d.clustering.map(e => toCatalyst(e, query, resolver))
+        case _: UnspecifiedDistribution =>
+          Array.empty[catalyst.expressions.Expression]
+      }
+
+      val queryWithDistribution = if (distribution.nonEmpty) {
+        val numShufflePartitions = sqlConf.numShufflePartitions
+        // repartition by expression will correctly pick range or hash partitioning
+        // based on whether we have an ordered or clustered distribution
+        // all expressions in OrderedDistribution will be mapped into SortOrder
+        // all expressions in ClusteredDistribution will be mapped into Expression
```

Comment:
This comment wasn't clear until I took a look at `RepartitionByExpression`. I think it may be more clear to say "the conversion to catalyst expressions above produces SortOrder for OrderedDistribution and Expression for ClusteredDistribution, and RepartitionByExpression will detect SortOrder expressions and use range partitioning."

Reply:
Fixed.